solr/src/main/scala/akka/stream/alpakka/solr/SolrFlowStage.scala

```diff
@@ -0,0 +1,184 @@
+/*
+ * Copyright (C) 2016-2018 Lightbend Inc. <http://www.lightbend.com>
+ */
+
+package akka.stream.alpakka.solr
+
+import java.net.{ConnectException, SocketException}
+
+import akka.NotUsed
+import akka.stream.alpakka.solr.SolrFlowStage.{Finished, Idle, Sending}
+import akka.stream.alpakka.solr.scaladsl.SolrSinkSettings
+import akka.stream.stage.{GraphStage, InHandler, OutHandler, TimerGraphStageLogic}
+import akka.stream.{Attributes, FlowShape, Inlet, Outlet}
+import org.apache.http.NoHttpResponseException
+import org.apache.solr.client.solrj.SolrClient
+import org.apache.solr.common.{SolrException, SolrInputDocument}
+
+import scala.collection.JavaConverters._
+import scala.collection.mutable
+import scala.concurrent.Future
+import scala.concurrent.duration._
+import scala.util.control.NonFatal
+
+object IncomingMessage {
+  // Apply method to use when not using passThrough
+  def apply[T](source: T): IncomingMessage[T, NotUsed] =
+    IncomingMessage(source, NotUsed)
+
+  // Java-api - without passThrough
+  def create[T](source: T): IncomingMessage[T, NotUsed] =
+    IncomingMessage(source)
+
+  // Java-api - with passThrough
+  def create[T, C](source: T, passThrough: C): IncomingMessage[T, C] =
+    IncomingMessage(source, passThrough)
+
+}
+
+final case class IncomingMessage[T, C](source: T, passThrough: C)
+
+final case class IncomingMessageResult[T, C](source: T, passThrough: C, status: Int)
+
+class SolrFlowStage[T, C](
```

Comment:
Please tighten the visibility with `private[solr]` to keep the internal APIs from being used.